### PR TITLE
Macro-based `Feature` and `Scenario` definition

### PR DESCRIPTION
**Pull request by OlegIlyenko**
This is just an idea with a small proof of concept, so I don't expect it to be merged :) but I hope it can be helpful. 

I was thinking that it can be to get rid of these `implicit` builders (described in #18) with macro-based solution. Here I just make a small macro-based rewrite of a DSL syntax:

```scala
Feature("OpenMovieDatabase API") {
  Scenario("search for GOT") {
    When I GET("/", params = "t" -> "Game of Thrones")

    Then assert status_is(200)
    
    // ...
  }
}
```

to a normal case class composition:

```scala
FeatureDef("OpenMovieDatabase API", Vector(
  Scenario("search for GOT", Vector(
    When I GET("/", params = "t" -> "Game of Thrones"),
    Then assert status_is(200)))))
```

It has some nice properties:

* no implicit arguments are nessesary
* no mutable builders are involved - DSL is completely immutable
* more control over the error-messages in the DSL. There is an opportunity to provide a helpful compilation errors

As a disadvantage, everything inside of `Scanario {...}` and `Feature {...}` need to be analyzed by the macro (only 1 level of AST). One can not declare variables in there, for instance (which can actually be an advantage). This can be implemented in the macro later though.

I prefixed all of the relevant parts of the DSL in order not to interfere with existing one.

Let me know what you think about it (first I wanted to comment on #18, but than I thought that actual code can describe the concept a bit better)

### CODE REVIEW

**Comment by agourlay** [comment]
Thank a lot for your work, the DSL looks great :trophy:

I will review this ASAP using my poor knowledge of macros, time to step up my game!

File: src/main/scala/com/github/agourlay/cornichon/dsl/MacroBasedDsl.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import com.github.agourlay.cornichon.core._
+import com.github.agourlay.cornichon.macros.Macro
+
+import scala.language.experimental.{ macros ⇒ `scalac, please just let me do it!` }
+
+trait MacroBasedDsl {
+  def FeatureMacro(name: String)(body: ⇒ Scenario): FeatureDef = macro Macro.featureImpl
+
+  def ScenarioMacro(name: String)(body: ⇒ Step): Scenario = macro Macro.scenarioImpl
+  def ScenarioMacro(name: String, ignore: Boolean)(body: ⇒ Step): Scenario = macro Macro.scenarioWithIgnoreImpl
```
Comment (agourlay) [inline src/main/scala/com/github/agourlay/cornichon/dsl/MacroBasedDsl.scala]: Macro do not play nice with params with default values?
I am fine with the overloaded version, it is just out of curiosity.

Comment (OlegIlyenko) [inline src/main/scala/com/github/agourlay/cornichon/dsl/MacroBasedDsl.scala]: yeah, as far as I can tell, named and default arguments are not supported yet. Once they were even announced, but didn't make into the scala release:

http://stackoverflow.com/a/34145464/576766

**Comment by agourlay** [comment]
I like this proof of concept and I think the library will greatly benefit from it :+1: 

So I propose to make the macro-based the default one and get rid of the mutable implicit one!

You can open another PR introducing the macro per default.

If you don't have time for this I could do it myselft as well although I like people people getting credit for their work in the git commit log :bowtie:

**Comment by OlegIlyenko** [comment]
Great, glad that you liked it! Sure, would be happy to prepare a proper PR :)

**Comment by OlegIlyenko** [comment]
@agourlay here we go. I made a macro-based DSL a default one. I also refactored the rest of the code to use this approach. (I squash the original commit into the new one)

I also changed the implementation of a macro quite a bit: it was not generic enough to handle other higher-kind step definitions (one that take steps as an argument) like `Repeat`, `Eventually`, `WithHeaders`, etc. The (black) magic is now in the `BodyElementCollector` (I don't like the name, but haven't come up with anything better :). It is very generic class that is used by other constructs like `Scenario`, `Feature`, `Repeat`, etc. It's also the only place where macro is used. The very nice thing about this is that it makes it possible for end-users to define custom higher-kind steps without creating any new macro: they can just use `BodyElementCollector`.

**Comment by agourlay** [comment]
Looks like xmas arrived a bit earlier this year :santa: 

Thank you for going the extra with the handling higher-kind steps definition, I will review this tomorrow.

**Comment by agourlay** [comment]
:rocket:

**Comment by agourlay** [comment]
fixes #18

**Pull request merged**